File: android/app/src/main/java/in/healthsoftcare/seniorcare/blufi/EspBlufiModule.kt
package `in`.healthsoftcare.seniorcare.blufi

import android.Manifest
import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothManager
import android.bluetooth.BluetoothProfile
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.os.Handler
import android.os.Looper
import androidx.core.content.ContextCompat
import blufi.espressif.BlufiCallback
import blufi.espressif.BlufiClient
import blufi.espressif.params.BlufiConfigureParams
import blufi.espressif.params.BlufiParameter
import blufi.espressif.response.BlufiScanResult
import blufi.espressif.response.BlufiStatusResponse
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import java.nio.charset.StandardCharsets

class EspBlufiModule(private val reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {

  companion object {
    private const val EVENT_DEVICE_FOUND = "EspBlufiDeviceFound"
    private const val EVENT_STATE = "EspBlufiState"
    private const val EVENT_WIFI_NETWORKS = "EspBlufiWifiNetworks"
    private const val EVENT_WIFI_STATUS = "EspBlufiWifiStatus"
    private const val SCAN_DURATION_MS = 12_000L
    private const val CONNECTION_TIMEOUT_MS = 15_000L
    private const val STATUS_RETRY_MS = 2_500L
    private const val MAX_STATUS_CHECKS = 5
  }

  private val mainHandler = Handler(Looper.getMainLooper())
  private val bluetoothAdapter: BluetoothAdapter? by lazy {
    val manager = reactContext.getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager
    manager?.adapter
  }
  private val discoveredDevices = linkedMapOf<String, BluetoothDevice>()
  private var scanning = false
  private var activeDeviceId: String? = null
  private var blufiClient: BlufiClient? = null
  private var secureSessionReady = false
  private var compatibilityMode = false
  private var provisioningSsid: String? = null
  private var statusCheckCount = 0

  override fun getName(): String = "EspBlufiModule"

  private fun emit(event: String, payload: WritableMap) {
    if (!reactContext.hasActiveReactInstance()) return
    reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit(event, payload)
  }

  private fun emitState(
    state: String,
    deviceId: String? = activeDeviceId,
    ssid: String? = null,
    message: String? = null,
  ) {
    val payload = Arguments.createMap()
    payload.putString("state", state)
    deviceId?.let { payload.putString("deviceId", it) }
    ssid?.let { payload.putString("ssid", it) }
    message?.let { payload.putString("message", it) }
    emit(EVENT_STATE, payload)
  }

  private fun hasPermission(permission: String): Boolean {
    return ContextCompat.checkSelfPermission(reactContext, permission) ==
      PackageManager.PERMISSION_GRANTED
  }

  private fun hasScanPermission(): Boolean {
    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
      hasPermission(Manifest.permission.BLUETOOTH_SCAN)
    } else {
      hasPermission(Manifest.permission.ACCESS_FINE_LOCATION)
    }
  }

  private fun hasConnectPermission(): Boolean {
    return Build.VERSION.SDK_INT < Build.VERSION_CODES.S ||
      hasPermission(Manifest.permission.BLUETOOTH_CONNECT)
  }

  @SuppressLint("MissingPermission")
  private fun stopNativeScan(emitStopped: Boolean) {
    if (!scanning) return
    bluetoothAdapter?.bluetoothLeScanner?.stopScan(scanCallback)
    mainHandler.removeCallbacks(stopScanRunnable)
    scanning = false
    if (emitStopped) {
      emitState("scanStopped", deviceId = null)
    }
  }

  private val stopScanRunnable = Runnable { stopNativeScan(emitStopped = true) }

  private val scanCallback = object : ScanCallback() {
    @SuppressLint("MissingPermission")
    override fun onScanResult(callbackType: Int, result: ScanResult) {
      val device = result.device
      discoveredDevices[device.address] = device

      val payload = Arguments.createMap()
      payload.putString("id", device.address)
      payload.putString(
        "name",
        result.scanRecord?.deviceName ?: device.name ?: "Nearby BLE device",
      )
      payload.putInt("rssi", result.rssi)
      payload.putBoolean(
        "isConnectable",
        Build.VERSION.SDK_INT < Build.VERSION_CODES.O || result.isConnectable,
      )
      val blufiServiceUuid = "0000ffff-0000-1000-8000-00805f9b34fb"
      val scanRecord = result.scanRecord
      val advertisesBluFi =
        scanRecord?.serviceUuids?.any {
          it.uuid.toString().equals(blufiServiceUuid, ignoreCase = true)
        } == true ||
          scanRecord?.serviceData?.keys?.any {
            it.uuid.toString().equals(blufiServiceUuid, ignoreCase = true)
          } == true
      if (!advertisesBluFi) return
      payload.putBoolean("isLikelyBluFi", true)
      emit(EVENT_DEVICE_FOUND, payload)
    }

    override fun onScanFailed(errorCode: Int) {
      scanning = false
      mainHandler.removeCallbacks(stopScanRunnable)
      emitState(
        "error",
        deviceId = null,
        message = "Bluetooth scan failed (code $errorCode).",
      )
    }
  }

  @ReactMethod
  fun startScan(promise: Promise) {
    if (!hasScanPermission() || !hasConnectPermission()) {
      promise.reject(
        "BLE_PERMISSION",
        "Bluetooth permission is required to scan for the pill dispenser.",
      )
      return
    }
    val adapter = bluetoothAdapter
    if (adapter == null || !adapter.isEnabled) {
      promise.reject("BLE_OFF", "Turn on Bluetooth to find the pill dispenser.")
      return
    }

    try {
      stopNativeScan(emitStopped = false)
      discoveredDevices.clear()
      val scanner = adapter.bluetoothLeScanner
      if (scanner == null) {
        promise.reject("BLE_UNAVAILABLE", "Bluetooth LE scanning is unavailable.")
        return
      }
      val settings =
        ScanSettings.Builder().setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY).build()
      scanner.startScan(null, settings, scanCallback)
      scanning = true
      emitState("scanning", deviceId = null)
      mainHandler.postDelayed(stopScanRunnable, SCAN_DURATION_MS)
      promise.resolve(true)
    } catch (error: Exception) {
      promise.reject("SCAN_FAILED", error.message, error)
    }
  }

  @ReactMethod
  fun stopScan(promise: Promise) {
    try {
      stopNativeScan(emitStopped = true)
      promise.resolve(true)
    } catch (error: Exception) {
      promise.reject("STOP_SCAN_FAILED", error.message, error)
    }
  }

  @SuppressLint("MissingPermission")
  @ReactMethod
  fun connect(deviceId: String, promise: Promise) {
    startConnection(deviceId, useCompatibilityMode = false, promise)
  }

  @SuppressLint("MissingPermission")
  @ReactMethod
  fun connectCompatibility(deviceId: String, promise: Promise) {
    startConnection(deviceId, useCompatibilityMode = true, promise)
  }

  @SuppressLint("MissingPermission")
  private fun startConnection(
    deviceId: String,
    useCompatibilityMode: Boolean,
    promise: Promise,
  ) {
    if (!hasConnectPermission()) {
      promise.reject(
        "BLE_PERMISSION",
        "Bluetooth permission is required to connect to the pill dispenser.",
      )
      return
    }
    val device = discoveredDevices[deviceId]
    if (device == null) {
      promise.reject("DEVICE_NOT_FOUND", "Scan again and select the pill dispenser.")
      return
    }

    try {
      stopNativeScan(emitStopped = false)
      closeClient()
      activeDeviceId = deviceId
      secureSessionReady = false
      compatibilityMode = useCompatibilityMode
      provisioningSsid = null
      statusCheckCount = 0

      val client = BlufiClient(reactContext.applicationContext, device)
      client.setGattWriteTimeout(10_000L)
      client.setGattCallback(gattCallback)
      client.setBlufiCallback(blufiCallback)
      blufiClient = client
      emitState("connecting", deviceId)
      client.connect()
      mainHandler.postDelayed(
        {
          if (blufiClient === client && !secureSessionReady) {
            closeClient()
            if (useCompatibilityMode) {
              emitState(
                "error",
                deviceId,
                message =
                  "The dispenser did not respond in BluFi compatibility mode. Its firmware may use a manufacturer-specific setup protocol.",
              )
            } else {
              emitState(
                "securityUnsupported",
                deviceId,
                message =
                  "The dispenser connected, but did not answer the standard encrypted BluFi handshake.",
              )
            }
          }
        },
        CONNECTION_TIMEOUT_MS,
      )
      promise.resolve(true)
    } catch (error: Exception) {
      emitState("error", deviceId, message = error.message ?: "Connection failed.")
      promise.reject("CONNECT_FAILED", error.message, error)
    }
  }

  @ReactMethod
  fun disconnect(promise: Promise) {
    try {
      closeClient()
      emitState("disconnected")
      activeDeviceId = null
      promise.resolve(true)
    } catch (error: Exception) {
      promise.reject("DISCONNECT_FAILED", error.message, error)
    }
  }

  @ReactMethod
  fun requestWifiScan(promise: Promise) {
    val client = readyClient(promise) ?: return
    client.requestDeviceWifiScan()
    promise.resolve(true)
  }

  @ReactMethod
  fun provision(ssid: String, password: String, promise: Promise) {
    val client = readyClient(promise) ?: return
    val normalizedSsid = ssid.trim()
    if (normalizedSsid.isEmpty()) {
      promise.reject("INVALID_SSID", "Select or enter a Wi-Fi network name.")
      return
    }

    val params = BlufiConfigureParams()
    params.setOpMode(BlufiParameter.OP_MODE_STA)
    params.setStaSSIDBytes(normalizedSsid.toByteArray(StandardCharsets.UTF_8))
    params.setStaPassword(password)
    provisioningSsid = normalizedSsid
    statusCheckCount = 0
    emitState("provisioning", ssid = normalizedSsid)
    client.configure(params)
    promise.resolve(true)
  }

  @ReactMethod
  fun requestWifiStatus(promise: Promise) {
    val client = blufiClient
    if (client == null) {
      promise.reject("NOT_CONNECTED", "Connect to the pill dispenser first.")
      return
    }
    client.requestDeviceStatus()
    promise.resolve(true)
  }

  private fun readyClient(promise: Promise): BlufiClient? {
    val client = blufiClient
    if (client == null) {
      promise.reject("NOT_CONNECTED", "Connect to the pill dispenser first.")
      return null
    }
    if (!secureSessionReady) {
      promise.reject("SESSION_NOT_READY", "The encrypted BluFi session is not ready.")
      return null
    }
    return client
  }

  private val gattCallback = object : BluetoothGattCallback() {
    @SuppressLint("MissingPermission")
    override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
      if (status == BluetoothGatt.GATT_SUCCESS &&
        newState == BluetoothProfile.STATE_CONNECTED
      ) {
        emitState("connected", gatt.device.address)
      } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
        secureSessionReady = false
        emitState("disconnected", gatt.device.address)
      } else if (status != BluetoothGatt.GATT_SUCCESS) {
        secureSessionReady = false
        emitState(
          "error",
          gatt.device.address,
          message = "Bluetooth connection failed (status $status).",
        )
      }
    }
  }

  private val blufiCallback = object : BlufiCallback() {
    override fun onGattPrepared(client: BlufiClient, status: Int, gatt: BluetoothGatt) {
      if (status != STATUS_SUCCESS) {
        emitState(
          "error",
          message = "Pill dispenser does not expose the expected BluFi service (code $status).",
        )
        return
      }
      emitState("connected")
      if (compatibilityMode) {
        secureSessionReady = true
        emitState("compatible")
      } else {
        client.negotiateSecurity()
      }
    }

    override fun onNegotiateSecurityResult(client: BlufiClient, status: Int) {
      if (status == STATUS_SUCCESS) {
        secureSessionReady = true
        emitState("secure")
      } else {
        secureSessionReady = false
        emitState(
          "securityUnsupported",
          message = "The dispenser rejected the standard encrypted BluFi handshake.",
        )
      }
    }

    override fun onDeviceScanResult(
      client: BlufiClient,
      status: Int,
      results: List<BlufiScanResult>,
    ) {
      if (status != STATUS_SUCCESS) {
        emitState("error", message = "The dispenser could not scan Wi-Fi (code $status).")
        return
      }
      val networks = Arguments.createArray()
      results.forEach { result ->
        if (result.ssid.isNullOrBlank()) return@forEach
        val network = Arguments.createMap()
        network.putString("ssid", result.ssid)
        network.putInt("rssi", result.rssi)
        networks.pushMap(network)
      }
      val payload = Arguments.createMap()
      payload.putArray("networks", networks)
      emit(EVENT_WIFI_NETWORKS, payload)
    }

    override fun onPostConfigureParams(client: BlufiClient, status: Int) {
      if (status == STATUS_SUCCESS) {
        emitState("configured", ssid = provisioningSsid)
        statusCheckCount = 0
        scheduleStatusCheck()
      } else {
        emitState(
          "error",
          ssid = provisioningSsid,
          message = "The dispenser rejected the Wi-Fi settings (code $status).",
        )
      }
    }

    override fun onDeviceStatusResponse(
      client: BlufiClient,
      status: Int,
      response: BlufiStatusResponse?,
    ) {
      if (status != STATUS_SUCCESS || response == null) {
        emitWifiStatus(
          connected = false,
          statusCode = status,
          message = "Could not read the dispenser's Wi-Fi status.",
        )
        return
      }

      val connected = response.isStaConnectWifi
      val ssid = response.staSSID ?: provisioningSsid
      emitWifiStatus(
        connected = connected,
        ssid = ssid,
        statusCode = response.staConnectionStatus,
        message =
          if (connected) {
            "Pill dispenser joined Wi-Fi."
          } else {
            "The dispenser is still connecting to Wi-Fi…"
          },
      )
      if (connected) {
        emitState("wifiConnected", ssid = ssid)
      } else if (provisioningSsid != null && statusCheckCount < MAX_STATUS_CHECKS) {
        scheduleStatusCheck()
      }
    }

    override fun onError(client: BlufiClient, errCode: Int) {
      emitState("error", message = "BluFi communication error (code $errCode).")
    }
  }

  private fun emitWifiStatus(
    connected: Boolean,
    ssid: String? = null,
    statusCode: Int? = null,
    message: String? = null,
  ) {
    val payload = Arguments.createMap()
    payload.putBoolean("connected", connected)
    ssid?.let { payload.putString("ssid", it) }
    statusCode?.let { payload.putInt("statusCode", it) }
    message?.let { payload.putString("message", it) }
    emit(EVENT_WIFI_STATUS, payload)
  }

  private fun scheduleStatusCheck() {
    statusCheckCount += 1
    mainHandler.postDelayed(
      {
        if (blufiClient != null && provisioningSsid != null) {
          blufiClient?.requestDeviceStatus()
        }
      },
      STATUS_RETRY_MS,
    )
  }

  private fun closeClient() {
    mainHandler.removeCallbacksAndMessages(null)
    stopNativeScan(emitStopped = false)
    try {
      blufiClient?.close()
    } finally {
      blufiClient = null
      secureSessionReady = false
      compatibilityMode = false
      provisioningSsid = null
      statusCheckCount = 0
    }
  }

  @ReactMethod
  fun addListener(eventName: String) {
    // Required by NativeEventEmitter.
  }

  @ReactMethod
  fun removeListeners(count: Int) {
    // Required by NativeEventEmitter.
  }

  override fun invalidate() {
    closeClient()
    super.invalidate()
  }
}

File: ios/EspBlufiBridge/EspBlufiModule.m
#import "EspBlufiModule.h"
#import <CoreBluetooth/CoreBluetooth.h>
#import <TargetConditionals.h>

#if !TARGET_OS_SIMULATOR
#import <EspBlufi/BlufiClient.h>
#endif

static NSString * const EspBlufiDeviceFoundEvent = @"EspBlufiDeviceFound";
static NSString * const EspBlufiStateEvent = @"EspBlufiState";
static NSString * const EspBlufiWifiNetworksEvent = @"EspBlufiWifiNetworks";
static NSString * const EspBlufiWifiStatusEvent = @"EspBlufiWifiStatus";
static NSTimeInterval const EspBlufiConnectionTimeoutSeconds = 15.0;

#if TARGET_OS_SIMULATOR
@interface EspBlufiModule () <CBCentralManagerDelegate>
#else
@interface EspBlufiModule () <CBCentralManagerDelegate, BlufiDelegate>
#endif
@property(nonatomic, assign) BOOL hasListeners;
@property(nonatomic, strong) CBCentralManager *scanManager;
@property(nonatomic, strong) NSMutableDictionary<NSString *, CBPeripheral *> *discoveredPeripherals;
@property(nonatomic, strong) NSString *activeDeviceId;
@property(nonatomic, strong) NSString *provisioningSsid;
@property(nonatomic, strong) NSString *blufiVersion;
@property(nonatomic, assign) BOOL secureSessionReady;
@property(nonatomic, assign) BOOL compatibilityMode;
@property(nonatomic, assign) NSInteger statusCheckCount;
#if !TARGET_OS_SIMULATOR
@property(nonatomic, strong) BlufiClient *client;
#endif
@end

@implementation EspBlufiModule

RCT_EXPORT_MODULE();

- (instancetype)init {
  self = [super init];
  if (self) {
    _discoveredPeripherals = [NSMutableDictionary new];
    _scanManager = [[CBCentralManager alloc] initWithDelegate:self
                                                        queue:dispatch_get_main_queue()];
  }
  return self;
}

+ (BOOL)requiresMainQueueSetup {
  return YES;
}

- (NSArray<NSString *> *)supportedEvents {
  return @[
    EspBlufiDeviceFoundEvent,
    EspBlufiStateEvent,
    EspBlufiWifiNetworksEvent,
    EspBlufiWifiStatusEvent
  ];
}

- (void)startObserving {
  self.hasListeners = YES;
}

- (void)stopObserving {
  self.hasListeners = NO;
}

- (void)emit:(NSString *)event body:(NSDictionary *)body {
  if (!self.hasListeners) return;
  [self sendEventWithName:event body:body];
}

- (void)emitState:(NSString *)state
         deviceId:(NSString *)deviceId
              ssid:(NSString *)ssid
           message:(NSString *)message {
  NSMutableDictionary *body = [NSMutableDictionary dictionaryWithObject:state forKey:@"state"];
  if (deviceId.length > 0) body[@"deviceId"] = deviceId;
  if (ssid.length > 0) body[@"ssid"] = ssid;
  if (message.length > 0) body[@"message"] = message;
  [self emit:EspBlufiStateEvent body:body];
}

- (void)rejectSimulator:(RCTPromiseRejectBlock)reject {
  reject(
    @"NOT_SUPPORTED",
    @"ESP-BluFi setup requires a physical iPhone because Bluetooth is unavailable in the iOS Simulator.",
    nil
  );
}

- (NSString *)bluetoothUnavailableReason {
  switch (self.scanManager.state) {
    case CBManagerStatePoweredOff:
      return @"Turn on Bluetooth to find the pill dispenser.";
    case CBManagerStateUnauthorized:
      return @"Bluetooth permission is required to find the pill dispenser.";
    case CBManagerStateUnsupported:
      return @"Bluetooth Low Energy is unsupported on this device.";
    case CBManagerStateResetting:
      return @"Bluetooth is resetting. Please try again.";
    case CBManagerStateUnknown:
    default:
      return @"Bluetooth is not ready yet. Please try again.";
  }
}

- (void)centralManagerDidUpdateState:(CBCentralManager *)central {
}

RCT_REMAP_METHOD(startScan,
                 startScanWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject) {
#if TARGET_OS_SIMULATOR
  [self rejectSimulator:reject];
#else
  if (self.scanManager.state != CBManagerStatePoweredOn) {
    reject(@"BLE_UNAVAILABLE", [self bluetoothUnavailableReason], nil);
    return;
  }
  [self.discoveredPeripherals removeAllObjects];
  [self.scanManager stopScan];
  [self.scanManager scanForPeripheralsWithServices:nil
                                           options:@{CBCentralManagerScanOptionAllowDuplicatesKey: @NO}];
  [self emitState:@"scanning" deviceId:nil ssid:nil message:nil];
  dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(12 * NSEC_PER_SEC)),
                 dispatch_get_main_queue(), ^{
    if (self.scanManager.isScanning) {
      [self.scanManager stopScan];
      [self emitState:@"scanStopped" deviceId:nil ssid:nil message:nil];
    }
  });
  resolve(@YES);
#endif
}

RCT_REMAP_METHOD(stopScan,
                 stopScanWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject) {
  [self.scanManager stopScan];
  [self emitState:@"scanStopped" deviceId:nil ssid:nil message:nil];
  resolve(@YES);
}

RCT_REMAP_METHOD(connect,
                 connectWithDeviceId:(NSString *)deviceId
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject) {
#if TARGET_OS_SIMULATOR
  [self rejectSimulator:reject];
#else
  [self startConnection:deviceId
      compatibilityMode:NO
                resolver:resolve
                rejecter:reject];
#endif
}

RCT_REMAP_METHOD(connectCompatibility,
                 connectCompatibilityWithDeviceId:(NSString *)deviceId
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject) {
#if TARGET_OS_SIMULATOR
  [self rejectSimulator:reject];
#else
  [self startConnection:deviceId
      compatibilityMode:YES
                resolver:resolve
                rejecter:reject];
#endif
}

#if !TARGET_OS_SIMULATOR
- (void)startConnection:(NSString *)deviceId
      compatibilityMode:(BOOL)compatibilityMode
                resolver:(RCTPromiseResolveBlock)resolve
                rejecter:(RCTPromiseRejectBlock)reject {
  if (self.discoveredPeripherals[deviceId] == nil) {
    reject(@"DEVICE_NOT_FOUND", @"Scan again and select the pill dispenser.", nil);
    return;
  }

  [self.scanManager stopScan];
  [self.client close];
  self.client = [[BlufiClient alloc] init];
  self.client.centralManagerDelete = self;
  self.client.blufiDelegate = self;
  self.activeDeviceId = deviceId;
  self.provisioningSsid = nil;
  self.blufiVersion = nil;
  self.secureSessionReady = NO;
  self.compatibilityMode = compatibilityMode;
  self.statusCheckCount = 0;
  [self emitState:@"connecting" deviceId:deviceId ssid:nil message:nil];
  [self.client connect:deviceId];
  NSString *attemptedDeviceId = [deviceId copy];
  BlufiClient *attemptedClient = self.client;
  dispatch_after(
    dispatch_time(
      DISPATCH_TIME_NOW,
      (int64_t)(EspBlufiConnectionTimeoutSeconds * NSEC_PER_SEC)
    ),
    dispatch_get_main_queue(),
    ^{
      if (self.client == attemptedClient &&
          [self.activeDeviceId isEqualToString:attemptedDeviceId] &&
          !self.secureSessionReady) {
        [self.client close];
        self.client = nil;
        NSString *versionSuffix = self.blufiVersion.length > 0
          ? [NSString stringWithFormat:@" (BluFi %@)", self.blufiVersion]
          : @"";
        if (compatibilityMode) {
          [self emitState:@"error"
                 deviceId:attemptedDeviceId
                      ssid:nil
                   message:
                     @"The dispenser did not respond in BluFi compatibility mode. Its firmware may use a manufacturer-specific setup protocol."];
        } else {
          [self emitState:@"securityUnsupported"
                 deviceId:attemptedDeviceId
                      ssid:nil
                   message:[NSString stringWithFormat:
                     @"The dispenser connected%@, but did not answer the standard encrypted BluFi handshake.",
                     versionSuffix]];
        }
      }
    }
  );
  resolve(@YES);
}
#endif

RCT_REMAP_METHOD(disconnect,
                 disconnectWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject) {
#if !TARGET_OS_SIMULATOR
  [self.client close];
  self.client = nil;
#endif
  [self.scanManager stopScan];
  self.secureSessionReady = NO;
  self.compatibilityMode = NO;
  self.provisioningSsid = nil;
  self.blufiVersion = nil;
  [self emitState:@"disconnected" deviceId:self.activeDeviceId ssid:nil message:nil];
  self.activeDeviceId = nil;
  resolve(@YES);
}

RCT_REMAP_METHOD(requestWifiScan,
                 requestWifiScanWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject) {
#if TARGET_OS_SIMULATOR
  [self rejectSimulator:reject];
#else
  if (![self requireSecureClient:reject]) return;
  [self.client requestDeviceScan];
  resolve(@YES);
#endif
}

RCT_REMAP_METHOD(provision,
                 provisionWithSsid:(NSString *)ssid
                 password:(NSString *)password
                 resolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject) {
#if TARGET_OS_SIMULATOR
  [self rejectSimulator:reject];
#else
  if (![self requireSecureClient:reject]) return;
  NSString *normalizedSsid =
    [ssid stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
  if (normalizedSsid.length == 0) {
    reject(@"INVALID_SSID", @"Select or enter a Wi-Fi network name.", nil);
    return;
  }

  BlufiConfigureParams *params = [BlufiConfigureParams new];
  params.opMode = OpModeSta;
  params.staSsid = normalizedSsid;
  params.staPassword = password ?: @"";
  self.provisioningSsid = normalizedSsid;
  self.statusCheckCount = 0;
  [self emitState:@"provisioning"
         deviceId:self.activeDeviceId
              ssid:normalizedSsid
           message:nil];
  [self.client configure:params];
  resolve(@YES);
#endif
}

RCT_REMAP_METHOD(requestWifiStatus,
                 requestWifiStatusWithResolver:(RCTPromiseResolveBlock)resolve
                 rejecter:(RCTPromiseRejectBlock)reject) {
#if TARGET_OS_SIMULATOR
  [self rejectSimulator:reject];
#else
  if (self.client == nil) {
    reject(@"NOT_CONNECTED", @"Connect to the pill dispenser first.", nil);
    return;
  }
  [self.client requestDeviceStatus];
  resolve(@YES);
#endif
}

#if !TARGET_OS_SIMULATOR
- (BOOL)requireSecureClient:(RCTPromiseRejectBlock)reject {
  if (self.client == nil) {
    reject(@"NOT_CONNECTED", @"Connect to the pill dispenser first.", nil);
    return NO;
  }
  if (!self.secureSessionReady) {
    reject(@"SESSION_NOT_READY", @"The encrypted BluFi session is not ready.", nil);
    return NO;
  }
  return YES;
}

- (void)scheduleStatusCheck {
  self.statusCheckCount += 1;
  dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2.5 * NSEC_PER_SEC)),
                 dispatch_get_main_queue(), ^{
    if (self.client != nil && self.provisioningSsid.length > 0) {
      [self.client requestDeviceStatus];
    }
  });
}

- (void)centralManager:(CBCentralManager *)central
 didDiscoverPeripheral:(CBPeripheral *)peripheral
     advertisementData:(NSDictionary<NSString *, id> *)advertisementData
                  RSSI:(NSNumber *)RSSI {
  NSString *name = advertisementData[CBAdvertisementDataLocalNameKey] ?: peripheral.name ?: @"";
  NSString *normalizedName = name.lowercaseString;
  NSArray<CBUUID *> *serviceUuids = advertisementData[CBAdvertisementDataServiceUUIDsKey] ?: @[];
  BOOL advertisesBluFi = [serviceUuids containsObject:[CBUUID UUIDWithString:@"FFFF"]];
  BOOL nameMatches =
    [normalizedName containsString:@"blufi"] ||
    [normalizedName containsString:@"pill"] ||
    [normalizedName containsString:@"dispenser"];
  NSNumber *connectable = advertisementData[CBAdvertisementDataIsConnectable] ?: @YES;

  NSString *identifier = peripheral.identifier.UUIDString;
  self.discoveredPeripherals[identifier] = peripheral;
  [self emit:EspBlufiDeviceFoundEvent
         body:@{
           @"id": identifier,
           @"name": name.length > 0 ? name : @"Nearby BLE device",
           @"rssi": RSSI ?: @(-127),
           @"isConnectable": connectable,
           @"isLikelyBluFi": @(advertisesBluFi || nameMatches)
         }];
}

- (void)centralManager:(CBCentralManager *)central
  didConnectPeripheral:(CBPeripheral *)peripheral {
  [self emitState:@"connected"
         deviceId:peripheral.identifier.UUIDString
              ssid:nil
           message:nil];
}

- (void)centralManager:(CBCentralManager *)central
didFailToConnectPeripheral:(CBPeripheral *)peripheral
                 error:(NSError *)error {
  [self emitState:@"error"
         deviceId:peripheral.identifier.UUIDString
              ssid:nil
           message:error.localizedDescription ?: @"Bluetooth connection failed."];
}

- (void)centralManager:(CBCentralManager *)central
didDisconnectPeripheral:(CBPeripheral *)peripheral
                 error:(NSError *)error {
  self.secureSessionReady = NO;
  [self emitState:@"disconnected"
         deviceId:peripheral.identifier.UUIDString
              ssid:nil
           message:error.localizedDescription];
}

- (void)blufi:(BlufiClient *)client
 gattPrepared:(BlufiStatusCode)status
      service:(CBService *)service
    writeChar:(CBCharacteristic *)writeChar
   notifyChar:(CBCharacteristic *)notifyChar {
  if (client != self.client) return;
  if (status != StatusSuccess) {
    [self emitState:@"error"
           deviceId:self.activeDeviceId
                ssid:nil
             message:@"The dispenser does not expose the expected BluFi service."];
    return;
  }
  [self emitState:@"connected" deviceId:self.activeDeviceId ssid:nil message:nil];
  if (self.compatibilityMode) {
    self.secureSessionReady = YES;
    [self emitState:@"compatible"
           deviceId:self.activeDeviceId
                ssid:nil
             message:nil];
  } else {
    [client negotiateSecurity];
  }
}

- (void)blufi:(BlufiClient *)client didNegotiateSecurity:(BlufiStatusCode)status {
  if (client != self.client) return;
  if (status == StatusSuccess) {
    self.secureSessionReady = YES;
    [self emitState:@"secure" deviceId:self.activeDeviceId ssid:nil message:nil];
  } else {
    self.secureSessionReady = NO;
    [self emitState:@"securityUnsupported"
           deviceId:self.activeDeviceId
                ssid:nil
             message:@"The dispenser rejected the standard encrypted BluFi handshake."];
  }
}

- (void)blufi:(BlufiClient *)client
didReceiveDeviceVersionResponse:(BlufiVersionResponse *)response
       status:(BlufiStatusCode)status {
  if (client != self.client) return;
  if (status == StatusSuccess && response != nil) {
    self.blufiVersion = [response getVersionString];
    NSLog(@"Healthsoft BluFi device protocol version: %@", self.blufiVersion);
  } else {
    NSLog(@"Healthsoft could not read BluFi device protocol version (code %ld)",
          (long)status);
  }
}

- (void)blufi:(BlufiClient *)client
didReceiveDeviceScanResponse:(NSArray<BlufiScanResponse *> *)scanResults
       status:(BlufiStatusCode)status {
  if (status != StatusSuccess) {
    [self emitState:@"error"
           deviceId:self.activeDeviceId
                ssid:nil
             message:@"The dispenser could not scan Wi-Fi networks."];
    return;
  }

  NSMutableArray *networks = [NSMutableArray new];
  for (BlufiScanResponse *response in scanResults) {
    if (response.ssid.length == 0) continue;
    [networks addObject:@{@"ssid": response.ssid, @"rssi": @(response.rssi)}];
  }
  [self emit:EspBlufiWifiNetworksEvent body:@{@"networks": networks}];
}

- (void)blufi:(BlufiClient *)client didPostConfigureParams:(BlufiStatusCode)status {
  if (status == StatusSuccess) {
    [self emitState:@"configured"
           deviceId:self.activeDeviceId
                ssid:self.provisioningSsid
             message:nil];
    self.statusCheckCount = 0;
    [self scheduleStatusCheck];
  } else {
    [self emitState:@"error"
           deviceId:self.activeDeviceId
                ssid:self.provisioningSsid
             message:[NSString stringWithFormat:
               @"The dispenser rejected the Wi-Fi settings (code %ld).",
               (long)status]];
  }
}

- (void)blufi:(BlufiClient *)client
didReceiveDeviceStatusResponse:(BlufiStatusResponse *)response
       status:(BlufiStatusCode)status {
  if (status != StatusSuccess || response == nil) {
    [self emit:EspBlufiWifiStatusEvent
           body:@{
             @"connected": @NO,
             @"statusCode": @(status),
             @"message": @"Could not read the dispenser's Wi-Fi status."
           }];
    return;
  }

  BOOL connected = response.isStaConnectWiFi;
  NSString *ssid = response.staSsid.length > 0 ? response.staSsid : self.provisioningSsid;
  NSMutableDictionary *body =
    [@{
      @"connected": @(connected),
      @"statusCode": @(response.staConnectionStatus),
      @"message": connected
        ? @"Pill dispenser joined Wi-Fi."
        : @"The dispenser is still connecting to Wi-Fi…"
    } mutableCopy];
  if (ssid.length > 0) body[@"ssid"] = ssid;
  [self emit:EspBlufiWifiStatusEvent body:body];

  if (connected) {
    [self emitState:@"wifiConnected"
           deviceId:self.activeDeviceId
                ssid:ssid
             message:nil];
  } else if (self.provisioningSsid.length > 0 && self.statusCheckCount < 5) {
    [self scheduleStatusCheck];
  }
}

- (void)blufi:(BlufiClient *)client didReceiveError:(NSInteger)errCode {
  [self emitState:@"error"
         deviceId:self.activeDeviceId
              ssid:nil
           message:[NSString stringWithFormat:
             @"BluFi communication error (code %ld).",
             (long)errCode]];
}
#endif

- (void)invalidate {
#if !TARGET_OS_SIMULATOR
  [self.client close];
  self.client = nil;
#endif
  [self.scanManager stopScan];
  self.scanManager.delegate = nil;
  self.scanManager = nil;
  [super invalidate];
}

@end
